R/conditions.R

```diff
@@ -0,0 +1,815 @@
+#'
+#'     # If no jump to with_restarts, return default value:
+#'     ""
+#'   })
+#' }
+#'
+#' # Normal value for valid input:
+#' fn(TRUE)
+#'
+#' # Default value for bad input:
+#' fn(FALSE)
+#'
+#' # Change the default value if you need an empty character vector by
+#' # defining an inplace handler that jumps to the restart. It has to be
+#' # inplace because thrown handlers jump to the place where they are
+#' # established before being executed, and the restart is not defined
+#' # anymore at that point:
+#' rst_handler <- inplace(function(c) rst_jump("rst_empty_chr"))
+#' with_handlers(fn(FALSE), default_empty_string = rst_handler)
+#'
+#' # You can use restarting() to create restarting handlers easily:
+#' with_handlers(fn(FALSE), default_empty_string = restarting("rst_null"))
+with_restarts <- function(.expr, ..., .restarts = list()) {
+  restarts <- c(list(...), .restarts)
+  with_restarts_(f_capture(.expr), restarts)
+}
+#' @rdname with_restarts
+#' @export
+with_restarts_ <- function(.expr, .restarts = list(), .env = NULL) {
+  f <- as_quoted_f(.expr, .env)
+  f <- env_set(f_interp(~withRestarts(uq(f), uqs(.restarts))), f)
+  f_eval(f)
+}
+
+
+# handlers -----------------------------------------------------------
+
+#' Create a thrown or in place handler.
+#'
+#' There are two types of condition handlers: exiting handlers, which
+#' are thrown to the place where they have been established (e.g.,
+#' \code{\link{with_handlers}()}'s evaluation frame), and local
+#' handlers, which are executed in place (e.g., where the condition
+#' has been signalled). \code{thrown()} and \code{inplace()} create
+#' handlers suitable for \code{\link{with_handlers}()}.
+#'
+#' A subtle point in the R language is that conditions are not thrown,
+#' handlers are. \code{\link[base]{tryCatch}()} and
+#' \code{\link{with_handlers}()} actually catch handlers rather than
+#' conditions. When a critical condition signalled with
+#' \code{\link[base]{stop}()} or \code{\link{panic}()}, R inspects the
+#' handler stack and looks for a handler that can deal with the
+#' condition. If it finds a throwable handler, it throws it to the
+#' function that established it (\code{\link{with_handlers}()}). If R
+#' finds an inplace handler, it executes it locally. Only if no
+#' handler can deal with the critical condition will R jump to
+#' top-level (see \code{\link{rst_abort}()}).
+#'
+#' @param handler A handler function that takes a condition as
+#'   argument.
+#' @param muffle Whether to muffle the condition after executing an
+#'   inplace handler. The signalling function must have established a
+#'   muffling restart. Otherwise, an error will be issued.
+#' @seealso \code{\link{with_handlers}()} for examples,
+#'   \code{\link{restarting}()} for another kind of inplace handler.
+#' @export
+thrown <- function(handler) {
+  structure(handler, class = c("thrown", "handler"))
+}
+#' @rdname thrown
+#' @export
+inplace <- function(handler, muffle = FALSE) {
```

Comment:
Since the handler functions only ever have one argument, should we provide a formula/expression helper?

```R
inplace(~ print(.))
thrown(~ 10)
```